src/filesystem_helper.cpp

```diff
@@ -0,0 +1,234 @@
+//      contributors may be used to endorse or promote products derived from
+//      this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+// This file is originally from:
+// https://github.com/ros/pluginlib/blob/1a4de29fa55173e9b897ca8ff57ebc88c047e0b3/pluginlib/include/pluginlib/impl/filesystem_helper.hpp
+
+/*! \file filesystem_helper.hpp
+ * \brief Cross-platform filesystem helper functions and additional emulation of [std::filesystem](https://en.cppreference.com/w/cpp/filesystem).
+ *
+ * Note: Once std::filesystem is supported on all ROS2 platforms, this class
+ * can be deprecated/removed in favor of the built-in functionality.
+ */
+
+#include "rcpputils/filesystem_helper.hpp"
+
+#include <cstring>
+
+#ifdef _WIN32
+#  define NOMINMAX
+#  define NOGDI
+#  include <windows.h>
+#  include <direct.h>
+#  include <fileapi.h>
+#  include <io.h>
+#  define access _access_s
+#else
+#  include <dirent.h>
+#  include <sys/types.h>
+#  include <unistd.h>
+#endif
+
+#include "rcutils/get_env.h"
+
+namespace rcpputils
+{
+namespace fs
+{
+
+bool path::exists() const
+{
+  return access(path_.c_str(), 0) == 0;
+}
+
+bool path::is_directory() const noexcept
+{
+  struct stat stat_buffer;
+  const auto rc = stat(path_.c_str(), &stat_buffer);
+
+  if (rc != 0) {
+    return false;
+  }
+
+#ifdef _WIN32
+  return (stat_buffer.st_mode & S_IFDIR) == S_IFDIR;
+#else
+  return S_ISDIR(stat_buffer.st_mode);
+#endif
+}
+
+bool path::is_regular_file() const noexcept
+{
+  struct stat stat_buffer;
+  const auto rc = stat(path_.c_str(), &stat_buffer);
+
+  if (rc != 0) {
+    return false;
+  }
+
+#ifdef _WIN32
+  return (stat_buffer.st_mode & S_IFREG) == S_IFREG;
+#else
+  return S_ISREG(stat_buffer.st_mode);
+#endif
+}
+
+path temp_directory_path()
+{
+#ifdef _WIN32
+#ifdef UNICODE
+#error "rcpputils::fs does not support Unicode paths"
+#endif
+  TCHAR temp_path[MAX_PATH];
+  DWORD size = GetTempPathA(MAX_PATH, temp_path);
+  if (size > MAX_PATH || size == 0) {
+    std::error_code ec(static_cast<int>(GetLastError()), std::system_category());
+    throw std::system_error(ec, "cannot get temporary directory path");
```

Comment:
Nit: `#include <system_error>` at the top for these symbols.

Reply:
done in https://github.com/ros2/rcpputils/pull/118/commits/b3280a4643004b2f81f6022b20bd02f733baa077